walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,128 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // If request duration is longer than a day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If no one has other events
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> mandatoryBusyTimes = new ArrayList<TimeRange>();
+    List<TimeRange> optionalBusyTimes = new ArrayList<TimeRange>();
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    
+    // Get mandatory attendees' busy times
+    if (request.getAttendees().size() > 0) {
+      mandatoryBusyTimes = getBusyTimes(events, request.getAttendees());
+      busyTimes.addAll(mandatoryBusyTimes);
+    }
+
+    // Get optional attendees' busy times
+    if (request.getOptionalAttendees().size() > 0) {
+      optionalBusyTimes = getBusyTimes(events, request.getOptionalAttendees());
+      busyTimes.addAll(optionalBusyTimes);
+    }
+    
+    // If there are busy times, find the open times that can fit the meeting duration
+    if (busyTimes.size() > 0) {
+      Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+      busyTimes = mergeOverlappedTimes(busyTimes);
+      List<TimeRange> openTimes = getOpenTimes(busyTimes, request.getDuration());
+      
+      // If there are no open times with the optional + mandatory, find open times for only the mandatory 
+      if (openTimes.size() > 0) {
+        return openTimes;
+      }
+      if (request.getAttendees().size() > 0) {
+        Collections.sort(mandatoryBusyTimes, TimeRange.ORDER_BY_START);
+        return getOpenTimes(mandatoryBusyTimes, request.getDuration());
+      }
+      // If there are no open times with the optional + mandatory and no mandatory attendees for the event,
+      return Arrays.asList();
+    }
+
+    // Return the whole day if there are no busy times
+    return Arrays.asList(TimeRange.WHOLE_DAY);
+  }
+
+  // Returns a list of busy times for "requestedAttendees", given other "events"
+  private List<TimeRange> getBusyTimes(Collection<Event> events, Collection<String> requestedAttendees) {
+
+    // Find events including requested attendees and add those time ranges to "busyTimes"
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    for (Event event: events) {
+      Set<String> attendees = event.getAttendees();
+
+      // Check if "event" includes a requested attendee
+      for (String requestedAttendee : requestedAttendees) {
+        if (attendees.contains(requestedAttendee)) {
+          busyTimes.add(event.getWhen());
+          break;
+        }
+      }
+    }
+    return busyTimes;
+  }
+  
+  // Return a new list of "busyTimes" with no overlapping time ranges
+  private List<TimeRange> mergeOverlappedTimes(List<TimeRange> busyTimes) {
+
+    List<TimeRange> busyTimesFinal = new ArrayList<TimeRange>();
+    TimeRange thisTime = busyTimes.get(0);
```

Comment:
Will there be an error if called with an empty list?

Reply:
Good catch. I added an early return